libs/cardano-ledger-test/src/Test/Cardano/Ledger/Constrained/Ast.hs

```diff
@@ -0,0 +1,453 @@
+{-# LANGUAGE AllowAmbiguousTypes #-}
+{-# LANGUAGE BangPatterns #-}
+{-# LANGUAGE FlexibleInstances #-}
+{-# LANGUAGE GADTs #-}
+{-# LANGUAGE PatternSynonyms #-}
+{-# LANGUAGE RankNTypes #-}
+{-# LANGUAGE StandaloneDeriving #-}
+{-# LANGUAGE TypeSynonymInstances #-}
+{-# LANGUAGE ViewPatterns #-}
+
+-- | The types that make up the Abstract Syntax Trees of the Language
+module Test.Cardano.Ledger.Constrained.Ast where
+
+import Cardano.Ledger.Coin (Coin (..), DeltaCoin (..))
+import Cardano.Ledger.Pretty
+import Data.Char (toLower)
+import qualified Data.List as List
+import Data.Map.Strict (Map)
+import qualified Data.Map.Strict as Map
+import Data.Set (Set)
+import qualified Data.Set as Set
+import Data.Text (Text, pack)
+import qualified Data.Universe as Univ (Any (..))
+import Data.Word (Word64)
+import Lens.Micro
+import Test.Cardano.Ledger.Constrained.Classes (Adds (..), Count (..), Sizeable (..), SumCond (..), Sums (..), runCond)
+import Test.Cardano.Ledger.Constrained.Combinators (errorMess)
+import Test.Cardano.Ledger.Constrained.Env (Access (..), AnyF (..), Env, Field (..), Name (..), V (..), findVar, storeVar)
+import Test.Cardano.Ledger.Constrained.Monad (Typed (..), failT)
+import Test.Cardano.Ledger.Constrained.TypeRep (Rep (..), Size (..), synopsis, testEql, (:~:) (Refl))
+
+-- ================================================
+
+data Literal era t where
+  Lit :: Rep era t -> t -> Literal era t
+
+data Term era t where
+  Fixed :: Literal era t -> Term era t
+  Var :: V era t -> Term era t
+  Dom :: Ord a => Term era (Map a b) -> Term era (Set a)
+  Rng :: (Ord a, Ord b) => Term era (Map a b) -> Term era (Set b)
+  ProjM :: (Ord a, Eq t) => Lens' b t -> Rep era t -> Term era (Map a b) -> Term era (Map a t)
+  ProjS :: (Ord b, Ord t) => Lens' b t -> Rep era t -> Term era (Set b) -> Term era (Set t)
+  Delta :: Term era Coin -> Term era DeltaCoin
+  Negate :: Term era DeltaCoin -> Term era DeltaCoin
+
+infix 4 :=:
+
+infix 4 :<=:
+
+data Pred era where
+  Sized :: Sizeable t => Term era Size -> Term era t -> Pred era
+  (:=:) :: Eq a => Term era a -> Term era a -> Pred era
+  (:<=:) :: Ord a => Term era (Set a) -> Term era (Set a) -> Pred era
```

Comment:
```suggestion
  (:~=:) :: Ord a => Term era (Set a) -> Term era (Set a) -> Pred era
```
Just a suggestion to avoid confusing this operator with a `<=` comparison-esque connotation. Nevermind if this, or, `(:-=:)` does not look any more useful to you. :slightly_smiling_face:

Reply:
I am thinking of using this :⊆:   instead

Reply:
In a way it is a comparison between sets similar to LTE, so I don't see a problem with this.

Reply:
I'd prefer to not have unicode operators